### ISSUE TITLE
omnisharp-roslyn cannot find Microsoft.Common.props

### ISSUE THREAD

**Issue by felschr** [NixOS/nixpkgs#126477]
**Describe the bug**
When running `omnisharp` I get an error that project loading failed due to a missing file:
```
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.MSBuild.ProjectManager","Message":"Loading project: /home/felschr/dev/work/project/Api/Api.csproj"},"Seq":147,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"WARNING","Name":"OmniSharp.MSBuild.ProjectManager","Message":"Failed to load project file '/home/felschr/dev/work/project/Api/Api.csproj'."},"Seq":148,"Type":"event"}
{"Event":"Error","Body":{"Text":"Microsoft.Build.Exceptions.InvalidProjectFileException: The imported project \"/nix/store/pw1ic8h00a9k2c5j1786ivq7zqrc59cb-omnisharp-roslyn-1.37.8/src/.msbuild/Current/Microsoft.Common.props\" was not found. Confirm that the expression in the Import declaration \"/nix/store/pw1ic8h00a9k2c5j1786ivq7zqrc59cb-omnisharp-roslyn-1.37.8/src/.msbuild/Current/Microsoft.Common.props\" is correct, and that the file exists on disk.  /nix/store/ycw6rph7kyff988mqk56qyf5q7qwidqc-dotnet-core-combined/sdk/5.0.202/Sdks/Microsoft.NET.Sdk/Sdk/Sdk.props\n  at Microsoft.Build.Shared.ProjectErrorUtilities.ThrowInvalidProject (System.String errorSubCategoryResourceName, Microsoft.Build.Shared.IElementLocation elementLocation, System.String resourceName, System.Object[] args) [0x00040] in <9be4d26643704d2e91efa63e9f29e6b0>:0 \n  at Microsoft.Build.Shared.ProjectErrorUtilities.VerifyThrowInvalidProject[T1,T2] (System.Boolean condition, System.String errorSubCategoryResourceName, Microsoft.Build.Shared.IElementLocation elementLocation, System.String resourceName, T1 arg0, T2 arg1) [0x00003] in <9be4d26643704d2e91efa63e9f29e6b0>:0
```

The path `/nix/store/pw1ic8h00a9k2c5j1786ivq7zqrc59cb-omnisharp-roslyn-1.37.8/src/.msbuild/Current/Microsoft.Common.props` does not seem to exist on my system.
There is only one folder in `/nix/store/pw1ic8h00a9k2c5j1786ivq7zqrc59cb-omnisharp-roslyn-1.37.8/src/.msbuild/Current`, which is called `Bin`.

**Expected behavior**
`omnisharp` should not throw an error and work as expected.

**Additional context**
`omnisharp.json`:
```json
{
  "MsBuild": {
    "UseLegacySdkResolver": true,
    "EnablePackageAutoRestore": true
  },
  "FormattingOptions": {
    "EnableEditorConfigSupport": true
  },
  "RoslynExtensionsOptions": {
    "EnableAnalyzersSupport": true
  }
}
```

Additionally, I've set the environment variables `DOTNET_ROOT`, `MSBuildSdksPath` & `MSBUILD_EXE_PATH` to the respective paths of my dotnet installation.

**Notify maintainers**
@tesq0 @ericdallo @corngood

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.12.7, NixOS, 21.11.20210606.fbfb794 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210601_5985b8b`
 - channels(root): `"nixos-21.11pre291991.ea7d4aa9b82"`
 - channels(felschr): `"nixpkgs-unstable-21.11pre291991.ea7d4aa9b82"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: omnisharp-roslyn
# a list of nixos modules affected by the problem
module:
```

**Comment by felschr** [comment]
I tried overriding the `installPhase` to copy the Microsoft.Common.props from omnisharp to the output (since I couldn't find it in the output of `msbuild`):

```nix
    installPhase = ''
      mkdir -p $out/bin
      cd ..
      cp -r src $out/
      rm -r $out/src/.msbuild
      cp -r ${msbuild}/lib/mono/msbuild $out/src/.msbuild

      chmod -R u+w $out/src
      mv $out/src/.msbuild/Current/{bin,Bin}
      
      # this line is new:
      cp -r src/.msbuild/Current/Microsoft.Common.props $out/src/.msbuild/Current

      makeWrapper ${mono}/bin/mono $out/bin/omnisharp \
      --add-flags "$out/src/OmniSharp.exe"
    '';
```

I don't get the original error anymore, but now `omnisharp` fails with `Method not found: string NuGet.Frameworks.NuGetFramework.get_Platform()`:
```
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.MSBuild.ProjectManager","Message":"Loading project: /home/felschr/dev/work/project/Api/Api.csproj"},"Seq":216,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"ERROR","Name":"OmniSharp.MSBuild.ProjectLoader","Message":"The \"ProcessFrameworkReferences\" task failed unexpectedly.\nSystem.MissingMethodException: Method not found: string NuGet.Frameworks.NuGetFramework.get_Platform()\n  at Microsoft.NET.Build.Tasks.ProcessFrameworkReferences.<ExecuteCore>b__105_1 (Microsoft.NET.Build.Tasks.ProcessFrameworkReferences+KnownFrameworkReference kfr) [0x00008] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at System.Linq.Enumerable+WhereEnumerableIterator`1[TSource].ToList () [0x0001b] in <9ba54d07696a449db4a4279fc05aa435>:0 \n  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <9ba54d07696a449db4a4279fc05aa435>:0 \n  at Microsoft.NET.Build.Tasks.ProcessFrameworkReferences.ExecuteCore () [0x00063] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at Microsoft.NET.Build.Tasks.TaskBase.Execute () [0x0002e] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x00029] in <355f0eeef3fc4a388d46eeb88cc51b2b>:0 \n  at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask (Microsoft.Build.BackEnd.ITaskExecutionHost taskExecutionHost, Microsoft.Build.BackEnd.Logging.TaskLoggingContext taskLoggingContext, Microsoft.Build.BackEnd.TaskHost taskHost, Microsoft.Build.BackEnd.ItemBucket bucket, Microsoft.Build.BackEnd.TaskExecutionMode howToExecuteTask) [0x002b9] in <355f0eeef3fc4a388d46eeb88cc51b2b>:0 "},"Seq":217,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"WARNING","Name":"OmniSharp.MSBuild.ProjectManager","Message":"Failed to load project file '/home/felschr/dev/work/project/Api/Api.csproj'."},"Seq":218,"Type":"event"}
{"Event":"MsBuildProjectDiagnostics","Body":{"FileName":"/home/felschr/dev/work/project/Api/Api.csproj","Warnings":[],"Errors":[{"LogLevel":"Error","FileName":"/nix/store/ycw6rph7kyff988mqk56qyf5q7qwidqc-dotnet-core-combined/sdk/5.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets","Text":"The \"ProcessFrameworkReferences\" task failed unexpectedly.\nSystem.MissingMethodException: Method not found: string NuGet.Frameworks.NuGetFramework.get_Platform()\n  at Microsoft.NET.Build.Tasks.ProcessFrameworkReferences.<ExecuteCore>b__105_1 (Microsoft.NET.Build.Tasks.ProcessFrameworkReferences+KnownFrameworkReference kfr) [0x00008] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at System.Linq.Enumerable+WhereEnumerableIterator`1[TSource].ToList () [0x0001b] in <9ba54d07696a449db4a4279fc05aa435>:0 \n  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <9ba54d07696a449db4a4279fc05aa435>:0 \n  at Microsoft.NET.Build.Tasks.ProcessFrameworkReferences.ExecuteCore () [0x00063] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at Microsoft.NET.Build.Tasks.TaskBase.Execute () [0x0002e] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x00029] in <355f0eeef3fc4a388d46eeb88cc51b2b>:0 \n  at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask (Microsoft.Build.BackEnd.ITaskExecutionHost taskExecutionHost, Microsoft.Build.BackEnd.Logging.TaskLoggingContext taskLoggingContext, Microsoft.Build.BackEnd.TaskHost taskHost, Microsoft.Build.BackEnd.ItemBucket bucket, Microsoft.Build.BackEnd.TaskExecutionMode howToExecuteTask) [0x002b9] in <355f0eeef3fc4a388d46eeb88cc51b2b>:0 ","StartLine":63,"StartColumn":5,"EndLine":0,"EndColumn":0}]},"Seq":219,"Type":"event"}
```

**Comment by corngood** [comment]
Could you share the project you're using to reproduce this?

Also, I'm locally using this change: https://github.com/NixOS/nixpkgs/pull/126487, which may be worth trying.

**Comment by felschr** [comment]
Ah, in the PR you're also copying `${msbuild}/lib/mono/xbuild/Current` to the output which contains the missing `Microsoft.Common.props`. Unfortunately, if I do the same in my `installPhase` override, I still get the `NuGetFramework.get_Platform()` not found error.

I can't share the entire project. Perhaps the project's `shell.nix` might help (reduced to the relevant parts):
```nix
let sources = import ./nix/sources.nix;

in { pkgs ? import sources.nixpkgs { } }:

with pkgs;
with lib;

let
  dotnet-combined = with dotnetCorePackages; combinePackages [ sdk_5_0 ];
  dotnetRoot = "${dotnet-combined}";
  dotnetSdk = "${dotnet-combined}/sdk";
  dotnetBinary = "${dotnetRoot}/bin/dotnet";

in mkShell {
  name = "env";
  buildInputs = [ dotnet-combined ];

  shellHook = ''
    export DOTNET_ROOT=${escapeShellArg dotnetRoot}
    local dotnetBase=${escapeShellArg dotnetSdk}/$(${
      escapeShellArg dotnetBinary
    } --version)
    export MSBuildSdksPath=$dotnetBase/Sdks
    export MSBUILD_EXE_PATH=$dotnetBase/MSBuild.dll
  '';
}
```

I'm not even sure if `omnisharp` picks up the `MSBuildSdksPath` environment variable.

**Comment by felschr** [comment]
I was able to reproduce this issue with a new .NET project created by `dotnet new console`.
Here's the project including the entire dotnet & omnisharp config:
https://github.com/felschr/omnisharp-nix-issue

**Comment by hpfr** [comment]
The PR mentioned before has been merged now; maybe that will help you @felschr ?

**Comment by felschr** [comment]
So, first of all I noticed that omnisharp was partially working before. The autocompletion seemed to mostly work for local variables & basic system libraries (e.g. `string.Compare`) but everything from NuGet wasn't autocomplete.

```
{"Event":"log","Body":{"LogLevel":"ERROR","Name":"OmniSharp.MSBuild.ProjectLoader","Message":"This project targets .NET 5.0 but the currently used MSBuild is not compatible with it - MSBuild 16.8+ is required. To
solve this, run OmniSharp on its embedded Mono (e.g. 'omnisharp.useGlobalMono':'never' in C# Extension for VS Code) or, if running on global Mono installation, make sure at least Mono 6.13 is installed on your mac
hine (https://mono-project.com/download/). Alternatively, add 'omnisharp.json' to your project root with the setting { \"msbuild\": { \"useBundledOnly\": true } }."},"Seq":225,"Type":"event"}
```

**Comment by felschr** [comment]
So, first of all I noticed that omnisharp was partially working before. The autocompletion seemed to mostly work for local variables & basic system libraries (e.g. `string.Compare`) but everything from NuGet wasn't autocomplete.

```
{"Event":"log","Body":{"LogLevel":"ERROR","Name":"OmniSharp.MSBuild.ProjectLoader","Message":"This project targets .NET 5.0 but the currently used MSBuild is not compatible with it - MSBuild 16.8+ is required. To
solve this, run OmniSharp on its embedded Mono (e.g. 'omnisharp.useGlobalMono':'never' in C# Extension for VS Code) or, if running on global Mono installation, make sure at least Mono 6.13 is installed on your mac
hine (https://mono-project.com/download/). Alternatively, add 'omnisharp.json' to your project root with the setting { \"msbuild\": { \"useBundledOnly\": true } }."},"Seq":225,"Type":"event"}
```

**Comment by felschr** [comment]
Adding `"msbuild": { "useBundledOnly": true }` to my `omnisharp.json` like suggested in the error message doesn't seem to have any effect. I'm still getting the same error.
Looks like mono is out of date, though. The error says it needs mono 6.13 while nixpkgs currently ships with 6.12.

**Comment by corngood** [comment]
We do have msbuild 16.8, but that doesn't seem to be sufficient.  `msbuild` in nixpkgs is the same thing that's used by `omnisharp-roslyn`, so I tried to build it:

```
➜  omnisharp-nix-issue git:(main) nix run nixpkgs.msbuild -c msbuild                                                                                                                                                                                                                               ~/src/omnisharp-nix-issue
Microsoft (R) Build Engine version 16.8.0-ci-21270-01 for Mono
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 30/07/2021 09:31:45.
Project "/home/david/src/omnisharp-nix-issue/omnisharp-nix-issue.csproj" on node 1 (default targets).
/nix/store/h9rgjz3ybkh24cvdkb8364ddl0ak6v0n-msbuild-16.8+xamarinxplat.2020.07.30.15.02/lib/mono/msbuild/Current/bin/Microsoft.Common.CurrentVersion.targets(1194,5): error MSB3644: The reference assemblies for .NETFramework,Version=v5.0 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks [/home/david/src/omnisharp-nix-issue/omnisharp-nix-issue.csproj]
```

I'm not sure about the reference assemblies.  We could try upgrading mono, but perhaps there's another way to acquire them?

**Comment by corngood** [comment]
I tried upgrading mono to 6.13 nightly, but it doesn't have .net 5 assemblies in `lib/mono/xbuild-frameworks`.

**Comment by felschr** [comment]
I thought `MSBuildSdksPath` is used to find those reference assemblies but it doesn't seem to work, at least in my setup.

So, I tried passing my specific dotnet package to `omnisharp-roslyn` to see if it can find the assemblies that way:
```nix
  omnisharp-roslyn =
    pkgs.omnisharp-roslyn.override { dotnet-sdk = dotnet-combined; };
```

This seems to allow omnisharp to find the SDK assemblies, but now it complains about `Method not found: string NuGet.Frameworks.NuGetFramework.get_Platform()` again.
```
{"Event":"log","Body":{"LogLevel":"ERROR","Name":"OmniSharp.MSBuild.ProjectLoader","Message":"The \"ProcessFrameworkReferences\" task failed unexpectedly.\nSystem.MissingMethodException: Method not found: string N
uGet.Frameworks.NuGetFramework.get_Platform()\n  at Microsoft.NET.Build.Tasks.ProcessFrameworkReferences.<ExecuteCore>b__105_1 (Microsoft.NET.Build.Tasks.ProcessFrameworkReferences+KnownFrameworkReference kfr) [0x
00008] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at System.Linq.Enumerable+WhereEnumerableIterator`1[TSource].ToList () [0x0001b] in <9ba54d07696a449db4a4279fc05aa435>:0 \n  at System.Linq.Enumerable.ToList[TSou
rce] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <9ba54d07696a449db4a4279fc05aa435>:0 \n  at Microsoft.NET.Build.Tasks.ProcessFrameworkReferences.ExecuteCore () [0x00063] in <c1a29c5daf134e4b
a6d73cbbae7974d1>:0 \n  at Microsoft.NET.Build.Tasks.TaskBase.Execute () [0x0002e] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHos
t.Execute () [0x00029] in <f3272346bd8f4581a546074d6bb7075f>:0 \n  at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask (Microsoft.Build.BackEnd.ITaskExecutionHost taskExecutionHost, Microsoft.Build.Back
End.Logging.TaskLoggingContext taskLoggingContext, Microsoft.Build.BackEnd.TaskHost taskHost, Microsoft.Build.BackEnd.ItemBucket bucket, Microsoft.Build.BackEnd.TaskExecutionMode howToExecuteTask) [0x002b9] in <f3
272346bd8f4581a546074d6bb7075f>:0 "},"Seq":21,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"WARNING","Name":"OmniSharp.MSBuild.ProjectManager","Message":"Failed to load project file '/home/felschr/dev/work/omnisharp-nix-issue/omnisharp-nix-issue.csproj'."},"Seq":22,"Typ
e":"event"}
{"Event":"MsBuildProjectDiagnostics","Body":{"FileName":"/home/felschr/dev/work/omnisharp-nix-issue/omnisharp-nix-issue.csproj","Warnings":[],"Errors":[{"LogLevel":"Error","FileName":"/nix/store/8qcd1d8lynh3b0a7z2
9xv5grfbm03slj-dotnet-core-combined/sdk/5.0.202/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets","Text":"The \"ProcessFrameworkReferences\" task failed unexpectedly.\nSystem.M
issingMethodException: Method not found: string NuGet.Frameworks.NuGetFramework.get_Platform()\n  at Microsoft.NET.Build.Tasks.ProcessFrameworkReferences.<ExecuteCore>b__105_1 (Microsoft.NET.Build.Tasks.ProcessFra
meworkReferences+KnownFrameworkReference kfr) [0x00008] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at System.Linq.Enumerable+WhereEnumerableIterator`1[TSource].ToList () [0x0001b] in <9ba54d07696a449db4a4279fc05a
a435>:0 \n  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <9ba54d07696a449db4a4279fc05aa435>:0 \n  at Microsoft.NET.Build.Tasks.ProcessFrameworkReferen
ces.ExecuteCore () [0x00063] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at Microsoft.NET.Build.Tasks.TaskBase.Execute () [0x0002e] in <c1a29c5daf134e4ba6d73cbbae7974d1>:0 \n  at Microsoft.Build.BackEnd.TaskExecut
ionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x00029] in <f3272346bd8f4581a546074d6bb7075f>:0 \n  at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask (Microsoft.Build.BackEnd.ITaskExec
utionHost taskExecutionHost, Microsoft.Build.BackEnd.Logging.TaskLoggingContext taskLoggingContext, Microsoft.Build.BackEnd.TaskHost taskHost, Microsoft.Build.BackEnd.ItemBucket bucket, Microsoft.Build.BackEnd.Tas
kExecutionMode howToExecuteTask) [0x002b9] in <f3272346bd8f4581a546074d6bb7075f>:0 ","StartLine":63,"StartColumn":5,"EndLine":0,"EndColumn":0}]},"Seq":23,"Type":"event"}
```

**Comment by corngood** [comment]
Yeah, there's some interaction between msbuild and the dotnet sdk that I don't fully understand.  Just investigating...

**Comment by felschr** [comment]
I think msbuild is supposed to find the reference assemblies from installed dotnet SDKs.
Makes sense to not bundle all assemblies with msbuild if only some versions are used.

But I also don't understand how it's trying to find those files and how omnisharp might interfere with that.
There are just much weird behaviour and so many options, I'm utterly confused by the entire thing.

I think to avoid having to override `omnisharp-roslyn` we should remove the `--prefix PATH : ${dotnet-combined}/bin \` part.
This also removes the need to install the current version of `dotnet-sdk`, since that might not be needed.

**Comment by corngood** [comment]
it was needed in $PATH because it expects the `dotnet` command to be available. 

https://github.com/OmniSharp/omnisharp-roslyn/pull/2058

**Comment by felschr** [comment]
I still think it should be up to the user to make sure it's in the PATH. Usually when installing dotnet it via nix-env, NixOS or home-manager that should happen automatically.
I don't think it makes much sense to use `omnisharp-roslyn` standalone, anyway.

**Comment by corngood** [comment]
I don't use the dotnet sdk at all for my project, so I don't want it in my environment, but without it in the PATH for omnisharp, omnisharp will currently fail.  If they support running omnisharp without it, I agree that it should be removed.

**Comment by felschr** [comment]
I'd be fine with using `pkgs.dotnet-sdk` as the default as long as it's possible to easily replace it either via arguments or by adding a different dotnet sdk to `PATH`. I'm not even sure why using `--prefix` prevents omnisharp to find my own dotnet installation.
Ideally, both of those options would be possible without needing to rebuild `omnisharp-roslyn`.

I've just found this https://github.com/dotnet/roslyn/issues/53906#issuecomment-875774866 regarding the NuGet error.
We might just have to update the `NuGet.Frameworks` nuget package to fix the error.
It seems to currently be at 5.7.0-rtm.6710: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/msbuild/deps.nix#L418
I tried updating that file but the script is complaining about a missing `NuGet.config`:
```
❯ ./create-deps.sh
{ fetchurl }: [
failed to load external entity "NuGet.config"
find: ‘.packages’: No such file or directory
find: ‘fake-home/.nuget/packages’: No such file or directory
```

**Comment by corngood** [comment]
I just pushed #132165 with some changes that seem to fix your test project, and still work on my project.

I still get a null-ref exception if I remove the $PATH entry for the sdk, but I upgraded it to .net 5, and I made it a suffix, which should allow it to be overridden.

I had to upgrade msbuild to fix the nuget conflicts between the dotnet sdk and msbuild, and then upgrade nuget in omnisharp-roslyn to the same version.
